Clusterwide K8s Cilium Network Policies

fristonio opened the pull request:
Fixes #9200

- [x] CRD for CiliumClusterwideNetworkPolicy
- [x] Controller for watching events related to CiliumClusterwideNetworkPolicy
- [x] RBAC registration for newly created CRD
- [x] Tests for the new policy CRD.
- [x] Add documentation for `CiliumClusterwideNetworkPolicy` resource.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9381)
<!-- Reviewable:end -->

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -269,6 +269,170 @@ type CiliumNetworkPolicyList struct {
+// CiliumGlobalNetworkPolicy is a Kubernetes third-party resource with an extended version
+// of CiliumNetworkPolicy which is cluster scoped rather than namespace scoped.
+type CiliumGlobalNetworkPolicy struct {
+	// +k8s:openapi-gen=false
```
If you do this we get all of the methods already implemented for CNP for free.
```
type CiliumGlobalNetworkPolicy struct {
    CiliumNetworkPolicy
}
```

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -269,6 +269,170 @@ type CiliumNetworkPolicyList struct {
 // +genclient
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
```
add this and generate the code again. This type of CiliumNetworkPolicy won't be namespaced.
```
// +genclient:nonNamespaced
```

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -269,6 +269,170 @@ type CiliumNetworkPolicyList struct {
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
```
same here for the non namespace tag

fristonio committed:
policy: add cilium clusterwide network policy crd and types

* Add CiliumGlobalNetowrkPolicy type to Cilium k8s api client.
* Update k8s utilities to incorporate changes corresponding to cilium
global policies.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
policy: add k8s controller to daemon for clusterwide policies

* Update k8s_watcher.go to include controller to watch for
CiliumClusterwideNetworkPolicy CRD.
* Update ExtractNamespace function in k8s utils to account for the fact
that any object with empty namespace is a cluster scoped object and we
should not automatically add default namespace to such objects.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
policy: add rbac rules for clusterwide network policy

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -266,6 +266,80 @@ type CiliumNetworkPolicyList struct {
+// rules. This is similar to the parse function used in CiliumNetworkPolicy with
+// the only difference that since the CRD is cluster scoped we don't send any namespace
+// here.
+func (r *CiliumClusterwideNetworkPolicy) Parse() (api.Rules, error) {
```
Was this removed in follow up commits?

aanm replied:
yes

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -266,6 +266,80 @@ type CiliumNetworkPolicyList struct {
+}
+
+// GetControllerName returns the unique name for the controller manager.
+func (r *CiliumClusterwideNetworkPolicy) GetControllerName() string {
```
Was this removed in follow up commits?

aanm replied:
yes

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -266,6 +266,80 @@ type CiliumNetworkPolicyList struct {
+}
+
+// GetIdentityLabels returns all rule labels in the CiliumNetworkPolicy.
+func (r *CiliumClusterwideNetworkPolicy) GetIdentityLabels() labels.LabelArray {
```
Was this removed in follow up commits?

aanm replied:
tes

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -276,58 +276,6 @@ type CiliumClusterwideNetworkPolicy struct {
 	CiliumNetworkPolicy
 }
 
-// Parse parses a CiliumClusterwideNetworkPolicy and returns a list of cilium policy
```
you can squash this change with the previous commit

aanm commented on pkg/k8s/factory_functions.go:
```diff
@@ -423,24 +439,54 @@ func ConvertToCNP(obj interface{}) interface{} {
-		cnp, ok := concreteObj.Obj.(*cilium_v2.CiliumNetworkPolicy)
-		if !ok {
-			return obj
+
```
Please create a new function dedicated for the conversion of `CiliumClusterwideNetworkPolicy` to `types.SlimpCNP`

aanm commented on pkg/k8s/factory_functions.go:
```diff
@@ -386,24 +387,39 @@ func ConvertToCNPWithStatus(obj interface{}) interface{} {
 		return &types.SlimCNP{
```
Please create a new function dedicated for the conversion of `CiliumClusterwideNetworkPolicy` to `types.SlimpCNP`

aanm reviewed (changes requested):
All good so far only small nits. Next step CI! :100:

fristonio committed:
test: add CI tests for Clusterwide policies

* Added k8s manifests for ccnp polices to test common policies
* Updated tests to include ccnp policy test manifests

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
policy: fix clusterwide policy status update issue for k8s<1.13

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

aanm commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -54,16 +58,24 @@ var (
-		labels.NewLabel(k8sConst.PolicyLabelUID, string(uid), labels.LabelSourceK8s),
 	}
+
+	if ns != "" {
```
add comment here:
```
// for clusterwide policies `ns` will be empty.
```

aanm commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -54,16 +58,24 @@ var (
 	}
+
+	if ns != "" {
+		labelsArr = append(labelsArr, labels.NewLabel(k8sConst.PolicyLabelNamespace, ns, labels.LabelSourceK8s))
```
to avoid long lines
```golang
nsLabel := labels.NewLabel(k8sConst.PolicyLabelNamespace, ns, labels.LabelSourceK8s)
labelsArr = append(labelsArr, nsLabel)
```

aanm commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -54,16 +58,24 @@ var (
+
+	labelsArr = append(labelsArr, labels.NewLabel(k8sConst.PolicyLabelUID, string(uid), labels.LabelSourceK8s))
+
+	return labelsArr
```
```golang

srcLabel := labels.NewLabel(k8sConst.PolicyLabelUID, string(uid), labels.LabelSourceK8s)

return append(labelsArr, srcLabel)
```

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -266,6 +273,32 @@ type CiliumNetworkPolicyList struct {
+type CiliumClusterwideNetworkPolicy struct {
+	CiliumNetworkPolicy
+
+	// Status is the status of the Cilium policy rule
```
```golang
	// Status is the status of the Cilium policy rule
	// +optional
	// The reason this field exists in this structure is due a bug in the k8s code-generator
	// that doesn't create a `StatusUpdate` method because the field does not exist in
        // the structure.
```

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -266,6 +273,32 @@ type CiliumNetworkPolicyList struct {
+// CiliumClusterwideNetworkPolicy is a Kubernetes third-party resource with an modified version
+// of CiliumNetworkPolicy which is cluster scoped rather than namespace scoped.
+type CiliumClusterwideNetworkPolicy struct {
+	CiliumNetworkPolicy
```
`*CiliumNetworkPolicy`

aanm commented on test/k8sT/manifests/ccnp-default-deny-egress.yaml:
```diff
@@ -0,0 +1,8 @@
+spec:
+  endpointSelector: {}
+  egress:
+  - {}
```
add new line

aanm commented on test/k8sT/manifests/ccnp-update-allow-all.yaml:
```diff
@@ -0,0 +1,13 @@
+    - {}
+  egress:
+  - toEndpoints:
+    - {}
```
add new line

aanm commented on test/k8sT/manifests/ccnp-update-allow-ingress.yaml:
```diff
@@ -0,0 +1,17 @@
+    toPorts:
+    - ports:
+      - port: "80"
+        protocol: TCP
```
add new line

aanm reviewed (changes requested):
Small nits but overall amazing accomplishment! :100:

aanm commented on pkg/k8s/factory_functions.go:
```diff
@@ -402,6 +436,52 @@ func ConvertToCNPWithStatus(obj interface{}) interface{} {
+// WARNING calling this function will set *all* fields of the given CNP as
+// empty.
+func ConvertToCCNP(obj interface{}) interface{} {
+	switch concreteObj := obj.(type) {
+	case *cilium_v2.CiliumClusterwideNetworkPolicy:
+		cnp := &types.SlimCNP{
+			CiliumNetworkPolicy: &cilium_v2.CiliumNetworkPolicy{
+				TypeMeta:   concreteObj.TypeMeta,
+				ObjectMeta: concreteObj.ObjectMeta,
+				Spec:       concreteObj.Spec,
+				Specs:      concreteObj.Specs,
+			},
+		}
+		*concreteObj = cilium_v2.CiliumClusterwideNetworkPolicy{}
+		return cnp
+
+	case cache.DeletedFinalStateUnknown:
+		cnp, ok := concreteObj.Obj.(*cilium_v2.CiliumClusterwideNetworkPolicy)
+		if !ok {
+			return obj
+		}
+		dfsu := cache.DeletedFinalStateUnknown{
+			Key: concreteObj.Key,
+			Obj: &types.SlimCNP{
+				CiliumNetworkPolicy: &cilium_v2.CiliumNetworkPolicy{
+					TypeMeta:   cnp.TypeMeta,
+					ObjectMeta: cnp.ObjectMeta,
+					Spec:       cnp.Spec,
+					Specs:      cnp.Specs,
+				},
+			},
+		}
+		*cnp = cilium_v2.CiliumClusterwideNetworkPolicy{}
+		return dfsu
+
+	default:
+		return obj
+	}
```
it looks to me that status is missing, no?

fristonio replied:
No, there are actually two methods `ConvertToCCNP` and `ConvertToCCNPWithStatus`. The first one converts the object to types.SlimCNP without the status field, therefore, there is no status here.

fristonio committed:
docs: add cilium clusterwide policy documentation

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

ianvernon commented on Documentation/kubernetes/policy.rst:
```diff
@@ -24,6 +24,11 @@ with Kubernetes:
   `CustomResourceDefinition` which supports specification of policies
   at Layers 3-7 for both ingress and egress.
 
+- The `CiliumClusterwideNetworkPolicy` format which is cluster scoped
```
a cluster-scoped

ianvernon commented on Documentation/kubernetes/policy.rst:
```diff
@@ -24,6 +24,11 @@ with Kubernetes:
   at Layers 3-7 for both ingress and egress.
 
+- The `CiliumClusterwideNetworkPolicy` format which is cluster scoped
+  `CustomResourceDefinition` for specifying cluster wide policies to be enforced
```
cluster-wide

ianvernon commented on Documentation/kubernetes/policy.rst:
```diff
@@ -99,3 +104,28 @@ Examples
+
+`CiliumClusterwideNetworkPolicy` is same as that of `CiliumNetworkPolicy` with the only
+difference in the scope of the policy. Policies defined by `CiliumClusterwideNetworkPolicy`
+are non namespaced and are cluster scoped. Internally the policy is composed of
```
non-namespaces and are cluster-scoped.

ianvernon commented on Documentation/policy/language.rst:
```diff
@@ -1228,3 +1228,25 @@ policies to a particular cluster.
+--------------------
+
+`CiliumNetworkPolicy` only allows to bind a policy restricted to a particular namespace. There can be situations
+where one wants to have a cluster scoped effect of the policy, this can be done using cilium's
```
cluster-scoped

ianvernon replied:
..., which can be done using Cilium's...

ianvernon commented on Documentation/policy/language.rst:
```diff
@@ -1228,3 +1228,25 @@ policies to a particular cluster.
+
+`CiliumNetworkPolicy` only allows to bind a policy restricted to a particular namespace. There can be situations
+where one wants to have a cluster scoped effect of the policy, this can be done using cilium's
+`CiliumClusterwideNetworkPolicy` kubernetes custom resource. The specification of the policy is same as that
```
kubernetes --> Kubernetes

ianvernon commented on Documentation/policy/language.rst:
```diff
@@ -1228,3 +1228,25 @@ policies to a particular cluster.
+`CiliumNetworkPolicy` only allows to bind a policy restricted to a particular namespace. There can be situations
+where one wants to have a cluster scoped effect of the policy, this can be done using cilium's
+`CiliumClusterwideNetworkPolicy` kubernetes custom resource. The specification of the policy is same as that
+of `CiliumNetworkPolicy` with the only difference that it is not namespaced.
```
with the only difference that it is not namespace --> except that it is not namespaced.

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -184,6 +193,52 @@ func createCNPCRD(clientset apiextensionsclient.Interface) error {
+		CustomResourceDefinitionPluralName = "ciliumclusterwidenetworkpolicies"
+
+		// CustomResourceDefinitionShortNames are the abbreviated names to refer to this CRD's instances
+		CustomResourceDefinitionShortNames = []string{"ccnp", "ciliumcnp"}
```
I don't think `ciliumcnp` is necessary here - CNP is almost always used to abbeviate `CiliumNetworkPolicy`, while here it refers to `ClusterwideNetworkPolicy`. I'd prefer we keep only `ccnp`.

ianvernon reviewed (commented):
Overall, looks great! I have a few nits mainly surrounding documentation grammar. Thanks for this contribution!!

tgraf commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -44,6 +44,10 @@ const (
+
+	// ResourceTypeCiliumClusterwideNetworkPolicy is the resource type used for the
+	// PolicyLabelDerivedFrom label
+	ResourceTypeCiliumClusterwideNetworkPolicy = "CiliumClusterwideNetworkPolicy"
```
I would call this `CiliumClusterNetworkPolicy`. It's shorter and also clear.

tgraf replied:
Discussed offline. The naming was discussed in SIG-policy and the conclusion was CiliumClusterwideNetworkPolicy. I'm OK with this.

tgraf reviewed (changes requested):
Fantastic work. The only change request I have is regarding the name. I think shorter is better but I'm also fine with the current proposal.
